File: 2021/Day4/Merge Intervals/Merge_intervals.cpp
class Solution//By CyFun
{
public:
    vector<vector<int>> merge(vector<vector<int>> &a)
    {
        if (a.size() == 1)
            return a;
        sort(a.begin(), a.end(), [](const vector<int> &v1, const vector<int> &v2) { return v1[0] < v2[0]; });

        vector<vector<int>> res;

        for (int i = 1; i < a.size(); i++)
        {
            if (a[i][0] <= a[i - 1][1])
            {
                a[i][0] = a[i - 1][0];
                a[i][1] = max(a[i][1], a[i - 1][1]);

                if (i == a.size() - 1)
                    res.push_back(a[i]);
            }
            else
            {
                res.push_back(a[i - 1]);

                if (i == a.size() - 1)
                    res.push_back(a[i]);
            }
        }

        return res;
    }
};

File: 2021/Day4/Merge Intervals/Merge_intervals.py
from collections import defaultdict
#By CyFun

class Solution(object):
    def __init__(self):
        self.startToEnds = defaultdict(list)

    def merge(self, intervals):
        """
        :type intervals: List[List[int]]
        :rtype: List[List[int]]
        """
        for (start, end) in intervals:
            self.startToEnds[start].append(end)

        intervals = [[start, max(endList)]
                     for start, endList in self.startToEnds.items()]
        intervals.sort(key=lambda x: x[0])

        mergedIntervals = []
        for interval in intervals:
            if not mergedIntervals or interval[0] > mergedIntervals[-1][1]:
                mergedIntervals.append(interval)
            else:
                mergedIntervals[-1][1] = max(mergedIntervals[-1]
                                             [1], interval[1])

        return mergedIntervals
